File: public/dsa/arrays/quick_sort/sol/quick_sort.ts
function partition(arr: number[], start: number, end: number): number {
  const pivot = arr[end];
  let i = start;

  for (let j = start; j < end; j++) {
    if (arr[j] < pivot) {
      [arr[i], arr[j]] = [arr[j], arr[i]];
      i++;
    }
  }

  [arr[i], arr[end]] = [arr[end], arr[i]];
  return i;
}

function quickSort(arr: number[], start: number, end: number): void {
  if (start < end) {
    const pivotIndex = partition(arr, start, end);
    quickSort(arr, start, pivotIndex - 1);
    quickSort(arr, pivotIndex + 1, end);
  }
}

function sortArray(nums: number[]): number[] {
  quickSort(nums, 0, nums.length - 1);
  return nums;
}

// tests-start
const nums = [5, 2, 3, 1];
console.log(sortArray(nums));
// tests-end

/*output
[1,2,3,5]
*/

export {};

File: public/dsa/arrays/quick_sort/sol/quick_sort.cpp
// imports-start
#include <bits/stdc++.h>
using namespace std;
// imports-end

int partition(vector<int>& arr, int start, int end) {
    int pivot = arr[end];
    int i = start;
    for (int j = start; j < end; ++j) {
        if (arr[j] < pivot) {
            swap(arr[i], arr[j]);
            ++i;
        }
    }
    swap(arr[i], arr[end]);
    return i;
}

void quick_sort(vector<int>& arr, int start, int end) {
    if (start < end) {
        int pivot_index = partition(arr, start, end);
        quick_sort(arr, start, pivot_index - 1);
        quick_sort(arr, pivot_index + 1, end);
    }
}

void quickSort(vector<int>& arr) {
    quick_sort(arr, 0, arr.size() - 1);
}

// tests-start
int main() {
    vector<int> nums = {5, 2, 3, 1};
    quickSort(nums);
    for (int n : nums) {
        cout << n << " ";
    }
    cout << endl;
    return 0;
}
// tests-end

/*output
1 2 3 5
*/